### DISCUSSION TITLE
Using HandleError::new() with extractors?

### DISCUSSION THREAD

**Discussion by cameronelliott** [tokio-rs/axum#1279] [Q&A] [answered]
I think Axum is awesome!

The new documentation on error handling is really good too: [error_handling](https://docs.rs/axum/latest/axum/error_handling/index.html)

One thing I cannot figure out, is whether HandleError::new() should work
with extractors or not.

```
    let app = Router::new()
        .route("/foo", post_service(HandleError::new(service, handle_anyhow_error)))
...

async fn handle(request: Request<Body>) -> Result<Response<Body>, anyhow::Error> {
    let b = Body::from("hello");
    Ok::<_, anyhow::Error>(Response::new(b))
}
```

Should I be able to add extractors to handle()?. I have been unable to get this to work.
If this is possible, it would be great to see it in the docs, or as a comment here.
(For newer Rust programmers myself)

Thanks

**Comment by davidpdrsn** [comment, marked as answer]
Yep that is possible. See https://docs.rs/axum/latest/axum/error_handling/index.html#running-extractors-for-error-handling

can you post more of your code including the error you get?

**Reply by cameronelliott** [reply]
Thanks for the prompt response. 💯 

I saw and studied that part of the docs also, but it just seems to demonstrate how to 
do an extractor with the handler for a layer/middleware.

I want to use an extractor not with the middleware error-handler, but
either before or inside the `thing_that_might_fail()`.

There is a minimal/not-working example here:
https://github.com/cameronelliott/axum-errors-extractor
**In the example, I want to use extractors with `thing_that_might_fail()`.**

I understand how to pass `Request<Body>` into `thing_that_might_fail()`, and I could
probably do everything the extractors do by-hand, but who wants to pull apart the `Request<Body>` by hand?

I'm sorry if I am missing something obvious.
Thank you for your help, I've been trying to get this going all day.

**Reply by davidpdrsn** [reply]
Oh so your question has nothing to do with `HandleError` then.

`thing_that_might_fail` is a _regular_ rust function that takes a `Path` and a `Request`. So you cannot just pass a `Request` to that. You can use [`Handler::into_service`](https://docs.rs/axum/latest/axum/handler/trait.Handler.html#method.into_service) to convert `thing_that_might_fail` (which implements `Handler`) into a `tower::Service` and then `call` it.

However I'm curious. Why are you even trying to do this in the first place? Why not use the regular `Handler`, `get`, `post`, etc. machinery that axum provides?

**Reply by cameronelliott** [reply]
David thank you.
Yes, good question, why don't I just use a handler?
Let's say I have this code:
`let app = Router::new().route("/foo", get(handler));`

Well, I have written a long function `handler` which
has maybe more than a dozen or more places where I want to use `...await?;` 
along with maybe `Ok("my body here")` at the end when there are no errors.

That's it, I just have a very long `Handler` function in which I need a lot of `...await?;`

Although as I am thinking about it, maybe I just need a regular Handler wrapping/calling an inner function
that returns a `Result<String, anyhow::Error>`, and then the handler converts it to a body.
**Maybe that's the best solution?**

What I was hoping to achieve was:
```
let app = Router::new().route("/foo", get(handler_xxx));

...

async handler_xxx(..extractors) -> Result<String,anyhow::Error> {
...
}
```

But now it seems that's not possible?
(Again, I understand I can just implement this as two functions:)
```
let app = Router::new().route("/foo", get(handler));

...
async handler(...extractors...) -> Result<String, StatusCode>  {
 let a = handler_inner(...);
 if Err(a) {
   // return status code
 }

async handler_xxx(..extractors) -> Result<String,anyhow::Error> {
   foo().await?;
   ...
}
```

Thanks for any more thoughts you can share.
Cameron

**Reply by davidpdrsn** [reply]
I recommend creating a wrapper around `anyhow::Error` that implements `IntoResponse` and `From<anyhow::Error>`:

```rust
use axum::{http::StatusCode, response::IntoResponse, routing::get, Router};
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", get(handler));

    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn handler() -> Result<(), AppError> {
    try_something().await?;

    Ok(())
}

async fn try_something() -> Result<(), anyhow::Error> {
    Ok(())
}

struct AppError(anyhow::Error);

impl From<anyhow::Error> for AppError {
    fn from(err: anyhow::Error) -> Self {
        AppError(err)
    }
}

impl IntoResponse for AppError {
    fn into_response(self) -> axum::response::Response {
        // how do you figure out what actually went wrong?! All the type information is erased in
        // `anyhow::Error`
        StatusCode::INTERNAL_SERVER_ERROR.into_response()
    }
}
```

But actually I don't recommend using `anyhow::Error` at all for web apps. The issue being that all type information is lost, so how do you convert the `anyhow::Error` into an appropriate response? I'd recommend `thiserror` instead so you actually know what went wrong and can create appropriate responses.

**Reply by cameronelliott** [reply]
Thank you so much @davidpdrsn I will try that tomorrow (past my bedtime). Very much appreciated. Cameron 💯
